### PR TITLE
[FLINK-32788][runtime] Support the handling of allocation errors when SpeculativeScheduler schedule speculative tasks by trigger fatal error.

### PR DESCRIPTION
**Pull request by JunRuiLee**
## What is the purpose of the change

Support the handling of allocation errors when SpeculativeScheduler schedule speculative tasks by trigger fatal error.

## Brief change log

Support the handling of allocation errors when SpeculativeScheduler schedule speculative tasks by trigger fatal error.

## Verifying this change

This change added tests and can be verified by SpeculativeSchedulerTest#testNotifySlowTaskWithAllocateSlotException

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no)**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no)**
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no)**
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by JunRuiLee** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/SpeculativeScheduler.java [hunk 1]
```diff
@@ -367,9 +373,13 @@ public void notifySlowTasks(Map<ExecutionVertexID, Collection<ExecutionAttemptID
             }
         }
 
-        executionDeployer.allocateSlotsAndDeploy(
-                newSpeculativeExecutions,
-                executionVertexVersioner.getExecutionVertexVersions(verticesToDeploy));
+        try {
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/SpeculativeScheduler.java:376]: This change does not fully solve the problem.
The underlying reason is that `SpeculativeScheduler#notifySlowTasks()` is not invoked from JobMaster RPC, but instead from a scheduled executor. e.g. if exception happens in `blockSlowNodes()`, similar problem can happen too.
Given that this kind of exceptions are unexpected, i.e. bugs, we can simply use `FutureUtils.assertNoException()` to handle it. A reference can be `DefaultScheduler#restartTasksWithDelay()`.

**Comment by JunRuiLee** [comment]
@zhuzhurk I'd updated this pr, PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/slowtaskdetector/ExecutionTimeBasedSlowTaskDetector.java [hunk 2]
```diff
@@ -94,6 +101,12 @@ public ExecutionTimeBasedSlowTaskDetector(Configuration configuration) {
                 this.baselineMultiplier);
     }
 
+    public ExecutionTimeBasedSlowTaskDetector(
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/slowtaskdetector/ExecutionTimeBasedSlowTaskDetector.java:104]: Should be annotated with `@VisibleForTesting`

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/slowtaskdetector/ExecutionTimeBasedSlowTaskDetector.java [hunk 3]
```diff
@@ -94,6 +101,12 @@ public ExecutionTimeBasedSlowTaskDetector(Configuration configuration) {
...
+            Configuration configuration, FatalErrorHandler fatalErrorHandler) {
+        this(configuration);
+        this.fatalErrorHandler = fatalErrorHandler;
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/slowtaskdetector/ExecutionTimeBasedSlowTaskDetector.java:107]: Better to checkNotNull

**Comment by JunRuiLee** [comment]
@zhuzhurk Thanks for the quick CR and I'd updated this pr, PTAL!

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Comment by JunRuiLee** [comment]
This PR has been blocked by https://issues.apache.org/jira/browse/FLINK-32876. I will wait for the resolution of https://github.com/apache/flink/pull/23222 and then rebase this PR.

**Comment by JunRuiLee** [comment]
I have rebased this PR, CI has been triggered again and it passed.

**Pull request closed without merging**